bip-taro.mediawiki

```diff
@@ -219,21 +225,139 @@ Taro-specific commitments, then the same or different leaf version can be used
+               case BranchReveal(sibling_1, sibling_2):
+                   expected_branch_hash = tagged_hash("TapBranch", sort(sibling_1, sibling_2))
+
+                   if expected_branch_hash != construct.sibling_hash:
+                       return false
+
+               # We know the tree is well formed now, so we'll verify the root
+               # commitment.
+               root_hash = tagged_hash("TapBranch", sort(taro_root_hash, control_block.sibling_hash))
+
+               expected_output_key = internal_key + tagged_hash("TapTweak", internal_key || root_hash)
+
+               return expected_output_key == witness_program(commitment_output.pk_script)
+
+       default:
+           return false
+
+</source>
+
+In addition to verifying that a commitment is unique within the constraints of
+a single taproot tree, we also need to verify that there is no similar
+commitment across the remaining outputs in the genesis transaction. We can do
+this by asserting that:
+* For each output on the genesis transaction that isn't creating an asset:
+** If top level key was derived using BIP 86, then verify the key derivation.
+** If the key commits to a script path:
+*** If only a single element is in the tree, verify that isn't a duplicate commitment.
+*** If the tree contains more than one element, then verify that the pre-image of the two branches isn't a taro commitment.
+
+The following algorithm can be used to verify that an output elsewhere in the
+transaction doesn't commit to a duplicate Taro commitment:
+<source lang="python">
+verify_no_taro_up_my_sleeves(genesis_tx: Tx, taro_root: []byte) -> bool
+
+    for tx_out in genesis_tx.tx_outs:
+        # We can ignore any non-P2TR output, or outputs we know have a valid
+        # commitment already.
+        if commitment_expected(tx_out):
+            continue
+
+        # We're expecting a commitment at this point, so we'll walk through
+        # our two cases.
+        match tx_out.taproot_commitment:
+            # A single leaf, so we just verify it isn't the taro commitment.
+            case SingleLeaf(leaf_bytes, internal_key):
+               # Someone has some 'splaining to do...
+               if leaf_bytes == taro_root:
+                   return false
+
+               leaf_hash = tagged_hash("TapLeaf", leaf_bytes)
+               expected_output_key = internal_key + tagged_hash("TapTweak", internal_key || leaf_hash)
+
+               if expected_output_key != witness_program(tx_out.pk_script):
+                   return false
+
+           # More than one element, so we just verify the opposite of what we
+           # expect for a unique commitment.
+           case MultiLeaf(branch_1, branch_2):
+              if taro_commitment_is_valid(tx_out, tx_out.ctrl_block, branch_1.reveal, taro_root):
+                  return false
+
+              if taro_commitment_is_valid(tx_out, tx_out.ctrl_block, branch_2.reveal, taro_root):
+                  return false
```

Comment:
Request for more elaboration here of how this satisfies the spec demands:
maybe:
`# More than one element, we need to verify that non of the other branches contain taro commitments`